File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.py
from collections import defaultdict

def solution(genres, plays):
    answer = []
    genre_total_playtimes = defaultdict(int)
    genre_playtimes = defaultdict(list)
    
    for i, (g, p) in enumerate(zip(genres, plays)):    # 장르별 가장 많이 재생된 노래를 찾기 위한 딕셔너리
        genre_total_playtimes[g] += p
        genre_playtimes[g].append((i, p))     # 장르별 노래의 고유번호, 플레이횟수
        
    for genre in sorted(genre_total_playtimes, key=lambda x:-genre_total_playtimes[x]):   # 장르별 총 재생횟수를 내림차순으로 정렬한 키값 리스트
        answer.extend([i for i, _ in sorted(genre_playtimes[genre], key=lambda x:(-x[1], x[0]))][:2])
    
    
    return answer

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.swift
import Foundation

func solution(_ genres:[String], _ plays:[Int]) -> [Int] {
    var totalDic: [String:Int] = [:]
    var playDic: [String: [(Int, Int)]] = [:]
    var answer:[Int] = []
    
    for (index, genre) in genres.enumerated() {
        playDic[genre, default:[]].append((index, plays[index]))
        totalDic[genre, default:0] += plays[index]
    }
    
    let sortedTotalDic = totalDic.sorted(by: { $0.1 > $1.1 })
    
    for (k, v) in sortedTotalDic {
        let sortedPlayDic = playDic[k]!.sorted(by: { $0.1 > $1.1 })
        if sortedPlayDic.count >= 2 {
            answer.append(sortedPlayDic[0].0)
            answer.append(sortedPlayDic[1].0)
        } else {
            answer.append(sortedPlayDic[0].0)
        }
    }
    
    return answer
}